develop/tutorials/articles/25-frontend-customizations/xx-converting-struts-action-wrappers-to-mvc-commands.markdown

```diff
@@ -1,60 +1,99 @@
-portal or portlet action respectively.
-
-You can see more information about it here:
-[https://dev.liferay.com/develop/tutorials/-/knowledge\_base/6-1/overriding-and-adding-struts-actions](https://www.google.com/url?q=https://dev.liferay.com/develop/tutorials/-/knowledge_base/6-1/overriding-and-adding-struts-actions&sa=D&ust=1465574286567000&usg=AFQjCNEWr1EsS_KdeKTRNSToDzaHkgQREQ)
-
-This mechanism no longer applies for most of the portal portlets
-starting on Liferay 7 because they are no longer using Struts Actions
-but MVCCommands. However, converting StrutsActionWrappers to MVCCommands
-is not complicated.
-
-Legacy StrutsActionWrappers would need to implement all the methods
-processAction, render, and serveResource even if we only needed to
-customize one of them. Each of this methods now can be customized
-independently using differnet class, making the logic simpler and easier
-to maintain. Depending on which method you customized, you will need to
-use a specific MVCCommand:
+# Converting StrutsActionWrappers to MVCCommands
+
+Since Liferay 6.1, developers could customize the Portal and Portlet Struts 
+Actions using a Hook and StrutsActionWrappers. For example, take a look at the 
+`liferay-hook.xml` file for a hook that overrode the login portlet's login 
+action:
+
+    <struts-action>
+        <struts-action-path>/login/login</struts-action-path>
+        <struts-action-impl>
+        com.liferay.sample.hook.action.ExampleStrutsPortletAction
+        </struts-action-impl>
+    </struts-action>
+    
+The `liferay-hook.xml` contains the struts mapping and the new class that 
+overrides the default login action.    
+
+The wrapper could extend either `BaseStrutsAction` or `BaseStrutsPortletAction`, 
+depending on whether the struts action was a portal or portlet action 
+respectively.
+
+Starting in Liferay 7.0 and Liferay DXP, this mechanism no longer applies for 
+most of the portal portlets because they are no longer using Struts Actions, but 
+instead use MVCCommands.
+
+This tutorial demonstrates the following:
+
+- How to convert your existing StrutsActionWrappers to MVCCommands
+
+Go ahead and get started.
+
+## Coverting Your old wrapper to MVCCommands
+
+Converting StrutsActionWrappers to MVCCommands is easier than you may think.
+
+As a review, legacy StrutsActionWrappers needed to implement all the methods,
+`processAction`, `render`, and `serveResource`, even if only one method was 
+being customized. Each of these methods can now be customized independently, 
+using different classes, making the logic simpler and easier to maintain. 
+Depending on which method you customized in your StrutsActionWrapper, you will 
+need to use the matching MVCCommand shown below:
 
 -   processAction ->  MVCActionCommand
 -   render -> MVCRenderCommand
 -   serveResource -> MVCResourceCommand
 
-Once you know which MVCCommand you will need to know which is the
-associated mapping that is used by that MVCCommand.
+<!--
```

Comment:
There's not much code to show here. I'm just trying to explain based on the customizations that the users could have what new java classes they should create new using the MVCCommand.

Take a look at the `ExampleStrutsPortletAction` class from here: https://dev.liferay.com/develop/tutorials/-/knowledge_base/6-2/overriding-and-adding-struts-actions . Depending on which actions were overriden the user will have to use different MVCCommands. In the example https://dev.liferay.com/develop/tutorials/-/knowledge_base/6-2/overriding-and-adding-struts-actions they only override the action and render so in order to migrate to the new pattern they would need to create 2 classes: MVCActionCommand and MVCRenderCommand.